swift/ql/test/library-tests/dataflow/flowsources/webview.swift

```diff
@@ -0,0 +1,43 @@
+
+// --- stubs ---
+class WKUserContentController {}
+class WKScriptMessage {}
+protocol WKScriptMessageHandler {
+    func userContentController(_ userContentController: WKUserContentController, didReceive message: WKScriptMessage)
+}
+protocol NSCopying {}
+protocol NSObjectProtocol {}
+class JSValue {}
+class JSContext {
+    var globalObject: JSValue { get { return JSValue() } }
+    func objectForKeyedSubscript(_: Any!) -> JSValue! { return JSValue() } 
+    func setObject(_: Any, forKeyedSubscript: (NSCopying & NSObjectProtocol) ) {}
```

Comment:
So to make sure I understand: in an early version you were looking for types that are handed to `setObject`, because JS can potentially taint them or anything contained in them.  But now you look for all types with the `JSExport` prototype because presumably that's required to do this, and is more accurate?

Reply:
Yes, I first thought that an object exposed through `setObject` would automatically export all its members to JS, but turns out it doesn't. You must declare explicitly, through a protocol that extends `JSExport`, the members that need to be exported. All the other members remain unaccessible from JavaScript.

And I thought that modeling the `JSExport` logic was enough. Theoretically the object would still need to be added to the `JSContext`, but I imagine that applications rarely would declare a type adopting a `JSExport`-inherited protocol and then not use it in the context. And still if they do, we probably want to alert if there's a dangerous path starting on them anyway, since it would be a ticking bomb waiting to explode when the object finally got added to the context.

But let me know if you think we should add the `setObject` constraint to the existing logic to be 100% accurate.

Reply:
This sounds quite reasonable as it is, if someone is marking something to export to JS but we can't see where, on balance I would probably prefer to still report a result.